bigbluebutton-html5/imports/ui/components/media/webcam-draggable-overlay/component.jsx

```diff
@@ -0,0 +1,510 @@
+
+import { styles } from '../styles';
+
+const propTypes = {
+  floatingOverlay: PropTypes.bool,
+  hideOverlay: PropTypes.bool,
+};
+
+const defaultProps = {
+  floatingOverlay: false,
+  hideOverlay: true,
+};
+
+export default class WebcamDraggableOverlay extends Component {
+  static getWebcamBySelector() {
+    return document.querySelector('video[class^="media"]');
+  }
+
+  static getIsOverlayChanged() {
+    const overlayToTop = document.querySelector('div[class*="overlayToTop"]');
+    const overlayToBottom = document.querySelector('div[class*="overlayToBottom"]');
+
+    return !!(overlayToTop || overlayToBottom);
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      dragging: false,
+      showDropZones: false,
+      showBgDropZoneTop: false,
+      showBgDropZoneBottom: false,
+      dropOnTop: false,
+      dropOnBottom: true,
+      initialPosition: { x: 0, y: 0 },
+      initialRectPosition: { x: 0, y: 0 },
+      lastPosition: { x: 0, y: 0 },
+      lastPositionInitSet: false,
+      resetPosition: false,
+      isFullScreen: false,
+    };
+
+    this.updateWebcamPositionByResize = this.updateWebcamPositionByResize.bind(this);
+
+    this.eventResizeListener = _.throttle(
+      this.updateWebcamPositionByResize,
+      500,
+      {
+        leading: true,
+        trailing: true,
+      },
+    );
+
+    this.handleWebcamDragStart = this.handleWebcamDragStart.bind(this);
+    this.handleWebcamDragStop = this.handleWebcamDragStop.bind(this);
+    this.handleFullscreenChange = this.handleFullscreenChange.bind(this);
+    this.fullscreenButtonChange = this.fullscreenButtonChange.bind(this);
+
+    this.setIsFullScreen = this.setIsFullScreen.bind(this);
+    this.setResetPosition = this.setResetPosition.bind(this);
+    this.setInitialReferencePoint = this.setInitialReferencePoint.bind(this);
+    this.setLastPosition = this.setLastPosition.bind(this);
+    this.setLastPositionInitSet = this.setLastPositionInitSet.bind(this);
+    this.setPositionAfterDropInEdge = this.setPositionAfterDropInEdge.bind(this);
+
+
+    this.getLastWebcamPosition = this.getLastWebcamPosition.bind(this);
+
+    this.dropZoneTopEnterHandler = this.dropZoneTopEnterHandler.bind(this);
+    this.dropZoneTopLeaveHandler = this.dropZoneTopLeaveHandler.bind(this);
+
+    this.dropZoneBottomEnterHandler = this.dropZoneBottomEnterHandler.bind(this);
+    this.dropZoneBottomLeaveHandler = this.dropZoneBottomLeaveHandler.bind(this);
+
+    this.dropZoneTopMouseUpHandler = this.dropZoneTopMouseUpHandler.bind(this);
+    this.dropZoneBottomMouseUpHandler = this.dropZoneBottomMouseUpHandler.bind(this);
+  }
+
+  componentDidMount() {
+    const { floatingOverlay } = this.props;
+    const { resetPosition } = this.state;
+
+    if (!floatingOverlay
+      && (!resetPosition)) this.setResetPosition(true);
+
+    window.addEventListener('resize', this.eventResizeListener);
+
+    const fullscreenChangedEvents = [
```

Comment:
Could we define these outside of the component (same level as propTypes) and reuse?